pkg/pillar/cmd/zedrouter/server.go

```diff
@@ -227,3 +253,238 @@ func (hdl hostnameHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	dirname, filename := path.Split(strings.TrimSuffix(r.URL.Path, "/"))
+	dirname = strings.TrimSuffix(dirname, "/")
+	remoteIP := net.ParseIP(strings.Split(r.RemoteAddr, ":")[0])
+	var hostname string
+	var id string
+	anStatus := lookupAppNetworkStatusByAppIP(hdl.ctx, remoteIP)
+	if anStatus != nil {
+		hostname = anStatus.DisplayName
+		id = anStatus.UUIDandVersion.UUID.String()
+	} else {
+		log.Errorf("No AppNetworkStatus for %s",
+			remoteIP.String())
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	switch filename {
+	case "meta-data":
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, "instance-id")
+		fmt.Fprintln(w, "hostname")
+		fmt.Fprintln(w, "public-keys/")
+		return
+	case "hostname":
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, hostname)
+		return
+	case "public-keys":
+		appConfig := lookupAppNetworkConfig(hdl.ctx, anStatus.Key())
+		if appConfig == nil {
+			log.Errorf("No AppNetworkConfig for %s",
+				anStatus.Key())
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		keys, err := getSSHPasswords(hdl.ctx, appConfig)
+		if err != nil {
+			log.Errorf("cannot get ssh passwords for %s: %v",
+				anStatus.Key(), err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+		for ind := range keys {
+			fmt.Fprint(w, fmt.Sprintf("%d=key-%d", ind, ind))
+		}
+		return
+	case "openssh-key":
+		_, previousFile := path.Split(dirname)
+		keyIndex, err := strconv.Atoi(previousFile)
+		if err == nil {
+			//request for ssh key content
+			appConfig := lookupAppNetworkConfig(hdl.ctx, anStatus.Key())
+			if appConfig == nil {
+				log.Errorf("No AppNetworkConfig for %s",
+					anStatus.Key())
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			keys, err := getSSHPasswords(hdl.ctx, appConfig)
+			if err != nil {
+				log.Errorf("cannot get ssh passwords for %s: %v",
+					anStatus.Key(), err)
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			if keyIndex < len(keys) {
+				w.WriteHeader(http.StatusOK)
+				fmt.Fprintln(w, keys[keyIndex])
+				return
+			}
+		}
+	case "instance-id":
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, id)
+		return
+	default:
+		if strings.Contains(r.URL.Path, "/public-keys/") {
+			keyIndex, err := strconv.Atoi(filename)
+			if err == nil {
+				//request for ssh key type
+				appConfig := lookupAppNetworkConfig(hdl.ctx, anStatus.Key())
+				if appConfig == nil {
+					log.Errorf("No AppNetworkConfig for %s",
+						anStatus.Key())
+					w.WriteHeader(http.StatusInternalServerError)
+					return
+				}
+				keys, err := getSSHPasswords(hdl.ctx, appConfig)
+				if err != nil {
+					log.Errorf("cannot get ssh passwords for %s",
+						anStatus.Key())
+					w.WriteHeader(http.StatusInternalServerError)
+					return
+				}
+				if keyIndex < len(keys) {
+					w.WriteHeader(http.StatusOK)
+					fmt.Fprint(w, "openssh-key")
+					return
+				}
+			}
```

Comment:
Shouldn't you return the public key somewhere here?

Reply:
Here we need to return type of key by number.